### PR TITLE
[apex] New rules for test method and assert statements

### PR DESCRIPTION
**Pull request by triandicAnt**
Before submitting a PR, please check that:
 - [x] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [x] `./mvnw clean verify` passes. This will [build](https://github.com/pmd/pmd/blob/master/BUILDING.md) and test PMD, execute PMD and checkstyle rules. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

**PR Description:**
pmd-apex new rules for best practices:
1. **System.assert** should have 2 param and **System.assertEquals/ System.assertNotEquals** should have 3 param.
```
   static void methodATest() {
        System.assertNotEquals('123', o.StageName); // not good
        System.assertEquals('123', o.StageName, o); // good
        System.assert(3 == var2); // not good
        System.assert(3 == var2, var2); // good
    }
```
2. test methods should have **@isTest** annotation.
```
    @isTest
    static void methodATest() { // good
    }
    static void methodBTest() { // not good
    }
    @isTest static void methodCTest() { // good
    }
```

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No java rules are changed!</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-apex/src/main/resources/category/apex/bestpractices.xml [hunk 1]
```diff
@@ -106,6 +106,56 @@ trigger Accounts on Account (before insert, before update, before delete, after
...
+]]>
+        </example>
+    </rule>
+    <rule name="ApexUnitTestMethodShouldHaveIsTestAnnotation"
+          since="6.12.0"
+          message="Apex test methods should have @isTest annotation."
+          class="net.sourceforge.pmd.lang.apex.rule.bestpractices.ApexUnitTestMethodShouldHaveIsTestAnnotationRule"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_apex_bestpractices.html#apexunittestmethodshouldhaveistestannotation">
+        <description>
+            Apex test methods should have @isTest annotation either in the same line or in previous line.
```
Comment (adangel) [inline pmd-apex/src/main/resources/category/apex/bestpractices.xml]: I'm not sure I understand the rule: Do all methods have to have the `@isTest` annotations? Or are there helper methods allowed in a test class?

> same line or previous line

Does that matter? At least in Java, the annotation always belongs to the following declaration regardless how much whitespace is in between. I don't know about Apex though...

Does this rule maybe want to solve the problem, that the `@isTest` method is maybe forgotten? Then this sounds like the new propose Java rule [DetachedTestCase](https://github.com/pmd/pmd/pull/1706).

**Comment by triandicAnt** [comment]
@adangel  Thank you for your feedback. I will be working on these feedback and will amend the changes to the pull request.

File: pmd-apex/src/main/resources/category/apex/bestpractices.xml (hunk 1 above)
Comment (triandicAnt) [reply pmd-apex/src/main/resources/category/apex/bestpractices.xml]: There is an existing PMD apex rule [ApexTestClassShouldHaveAssertsRule](https://pmd.github.io/latest/pmd_rules_apex_bestpractices.html#apexunittestclassshouldhaveasserts) that checks if a test class/method has an assert statement or not. This rule is kind of opposite to that. For example if there is a test method having assert statements then it should have `@isTest` annotation. (as [testMethod](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_classes_annotation_isTest.htm) has been deprecated and Salesforce advices to use `@isTest` annotation for test class/method, so I thought of adding this rule. However the helper methods in the test class need not have `@isTest` annotation.  
```
Does that matter?
```
No, It can have multiple new line spaces, but ideally there should not be more than one new line.

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            methodLocMap.put(testMethod.getNode().getLoc().getLine(), testMethod);
+        }
+        List<ASTModifierNode> modifierList = node.findDescendantsOfType(ASTModifierNode.class);
+        final Map<Integer, ModifierOrAnnotation> modifierLocMap = new HashMap<>();
+        for (ASTModifierNode modifier : modifierList) {
+            for (final ModifierOrAnnotationTypeInfo modifierOrAnnotationTypeInfo : modifier.getNode().getModifiers().all()) {
+                ModifierOrAnnotation modifierOrAnnotation = modifier.getNode().getModifiers().get(modifierOrAnnotationTypeInfo);
+                if (modifierOrAnnotation instanceof Annotation && TypeInfoEquivalence
+                        .isEquivalent(modifierOrAnnotationTypeInfo, AnnotationTypeInfos.IS_TEST)) {
+                    modifierLocMap.put(modifierOrAnnotation.getLoc().getLine(), modifierOrAnnotation);
```
Comment (adangel) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java]: Since the location of the annoation (the line number) is irrelevant, we should not check this at all.

Consider this code snippet:

```
@isTest
private class ATest {
    @isTest
    @SuppressWarnings
    static void methodATest() {
    }
    static void methodBTest() {
    }
    @isTest static void methodCTest() {
    }
    private void fetchData() {
    }
}
```

Since the annotation is not directly on the line before, we don't see, that `methodATest` is actually annotated.

If you look at the structure of the AST:
grafik

It should be enough to just check for the existence of the "IsTest" annoation. 

Now I understand your approach: any method having "test" in its name and that is in a class, that is annotated with "IsTest", should be annotated.

So, first step: collecting the methods:

```java
    private Object checkForIsTestAnnotation(final ApexNode<?> node, final Object data) {
        final List<ASTMethod> methods = node.findDescendantsOfType(ASTMethod.class);
        final List<ASTMethod> testMethods = new ArrayList<>();
        for (ASTMethod method : methods) {
            final Version classApiVersion = method.getNode().getDefiningType().getCodeUnitDetails().getVersion();
            if (!isTestMethodOrClass(method) && classApiVersion.isGreaterThan(Version.V174)
                    && method.getImage().toLowerCase(Locale.ROOT).contains(TEST)) {
                testMethods.add(method);
            }
        }
```

second step: reporting the violations:
```java
        for (ASTMethod method : testMethods) {
            addViolationWithMessage(data, node,
                  "''{0}'' method should have @IsTest annotation.",
                  new Object[] { method.getImage() });
        }
        return data;
    }
```

No need to check for line numbers.

I noticed now, that the rule can easily be written as a XPath expression:
```
//UserClass[ModifierNode/Annotation[@Image = 'IsTest']]
  /Method
    [matches(@Image, 'test', 'i')]
    [not(ModifierNode/Annotation[@Image = 'IsTest'])]
```

See https://pmd.github.io/pmd-6.12.0/pmd_userdocs_extending_writing_xpath_rules.html .
Maybe it's worth to rewrite this rule as a XPath rule? (the test cases stay the same, just the java class ApexUnitTestMethodShouldHaveIsTestAnnotationRule is not necessary and the entry in bestpractices.xml changes)

Comment (adangel) [reply pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java]: Caveat: the classApiVersion is not checked in XPath, so that might be a reason, why we need to write this for now as a Java rule

**Comment by rsoesemann** [comment]
Sorry for being late but I am absolutely against this PR. This is not at all a best practice. Sorry to say that but it’s just plain nonsense to mandate each assert equal to have an explaining string. Concise test with a name that reveals their intent and there is only one assert it’s just not needed. Not ever practice that works for you must be for all the others. @adangel #veto

**Comment by adangel** [comment]
> Not every practice that works for you must be for all the others.

Exactly, that's the point. I understand, that this rule "ApexUnitTestAssertStatement" is controversal. It probably cannot be used by everyone. But does it need to be? Selecting the rules, that fit the project's need is part of using PMD - blindly executing all rules is not a good idea...
See also [Best Practices](https://pmd.github.io/pmd-6.12.0/pmd_userdocs_best_practices.html): "Choose the rules that are right for you".

So, how about a compromise: The ruleset "rulesets/apex/ruleset.xml" seems to be some kind of default ruleset. Not sure, how this ruleset is used, but it seems to be used for code climate?

It would make sense then to NOT add the rule "ApexUnitTestAssertStatement" to this ruleset, since "it does not work for all users". The rule could still be used by users, who want to use it.

Now we've discussed the first rule, but this PR also contains a second rule: "ApexUnitTestMethodShouldHaveIsTestAnnotation". What about this one? It reminds me of the upcoming Java rule DetachedTestCase (see #1706)

**Comment by rsoesemann** [comment]
I‘d go with this rule is it’s a lean xpath rule and not something we later have to maintain. Should be a one liner. I agree with your view that we should allow somewhat controversial stuff which this is but also has some use cases.

**Comment by rsoesemann** [comment]
@adangel the is no codeclimate relation anymore in this repo nor the default ruleset. Besides that I just want to keep it simple and have a default uncurated ruleset.

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java (hunk 2 above)
Comment (rsoesemann) [reply pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java]: @triandicAnt why is `classApiVersion` important? I agree with @adangel that this rule could be written as a few line XPath rule opposed to a complex multi hundred line Java rule.

If there is no severe reason for checking the API version please rewrite the rule as lean XPath rule.

Comment (adangel) [reply pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java]: We check the "classApiVersion" also in the base test class `AbstractApexUnitTestRule`:
https://github.com/pmd/pmd/blob/5d86217871f086f8223da327099d1d67a6e45dab/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/AbstractApexUnitTestRule.java#L33

I assume, that the code of one class is always the same version. This means, we don't need to do this check again for the methods.

However, if we switch to XPath rule and want to have exactly the same behavior, then need to expose this version in e.g. ASTUserClass via

```java
    public double getVersion() {
        return node.getDefiningType().getCodeUnitDetails().getVersion().getExternal();
    }
```

And then write the XPath as follows:

```
//UserClass[ModifierNode/Annotation[@Image = 'IsTest']][@Version > 23.0]
  /Method
    [matches(@Image, 'test', 'i')]
    [not(ModifierNode/Annotation[@Image = 'IsTest'])]
```

However, maybe we don't need the version check at all? I looked up the mapping between the different version numbers apex uses (I finally found the mapping). The mapping is in the enum `apex.jorje.services.Version`:

```
V140 = external: 6.0 internal: 140.0
V142 = external: 7.0 internal: 142.0
V144 = external: 8.0 internal: 144.0
V146 = external: 9.0 internal: 146.0
V148 = external: 10.0 internal: 148.0
V150 = external: 11.0 internal: 150.0
V150_1 = external: 11.1 internal: 150.1
V152 = external: 12.0 internal: 152.0
V154 = external: 13.0 internal: 154.0
V156 = external: 14.0 internal: 156.0
V158 = external: 15.0 internal: 158.0
V160 = external: 16.0 internal: 160.0
V162 = external: 17.0 internal: 162.0
V164 = external: 18.0 internal: 164.0
V166 = external: 19.0 internal: 166.0
V168 = external: 20.0 internal: 168.0
V170 = external: 21.0 internal: 170.0
V172 = external: 22.0 internal: 172.0
V174 = external: 23.0 internal: 174.0
V176 = external: 24.0 internal: 176.0
V178 = external: 25.0 internal: 178.0
V180 = external: 26.0 internal: 180.0
V182 = external: 27.0 internal: 182.0
V184 = external: 28.0 internal: 184.0
V186 = external: 29.0 internal: 186.0
V188 = external: 30.0 internal: 188.0
V190 = external: 31.0 internal: 190.0
V192 = external: 32.0 internal: 192.0
V194 = external: 33.0 internal: 194.0
V196 = external: 34.0 internal: 196.0
V198 = external: 35.0 internal: 198.0
V200 = external: 36.0 internal: 200.0
V202 = external: 37.0 internal: 202.0
V204 = external: 38.0 internal: 204.0
V206 = external: 39.0 internal: 206.0
V208 = external: 40.0 internal: 208.0
V210 = external: 41.0 internal: 210.0
V212 = external: 42.0 internal: 212.0
V214 = external: 43.0 internal: 214.0
V216 = external: 44.0 internal: 216.0
V218 = external: 45.0 internal: 218.0
V220 = external: 46.0 internal: 220.0
V222 = external: 47.0 internal: 222.0
current: V220
compiler: V210
max: V220
min: V140
post_release: V212
```

We check against the internal version "174", which corresponds to external version "23.0". Now we are at version 45.0.
The check seems only to be relevant is someone still uses version 23.0... which seems to be a apex version 7 years old... (maybe that's the time, when isTest was introduced?).
I'm also not sure, whether the check works as we wanted: If I remove all the isTest annotations from the code, Jorje still parses it as classApiVersion=46.0 ... (which is actually the upcoming Summer 2019 release of apex).
Maybe we are using the Jorje Parser the wrong way?

(Another crazy idea: If we can correctly determine the api version of a apex compilation unit, we could leverage the minimumLanguageVersion/maximumLanguageVersion properties of the rules maybe?)

In short, what we could do in any case:
* Expose the version, so that it is available via XPath
* Write the rule ApexUnitTestMethodShouldHaveIsTestAnnotation as a XPath rule
* Figure out the version uncertainty later

Comment (rsoesemann) [reply pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java]: Sounds awesome...

**Comment by adangel** [comment]
Hi @triandicAnt ! Do you think you have time to continue on this PR? Or should we take over? Just let us know, thanks!

**Comment by triandicAnt** [comment]
@adangel  I was not able to install `pmd-designer` so was not able to test the xml rule. It was always throwing this error:
`You seem to be missing the JavaFX runtime. Please install JavaFX on your system and try again.`
even if I have $JAVAFX_HOME setup in my system:
`echo $JAVAFX_HOME
/Users/sudhansu/Downloads/javafx-sdk-11.0.2/lib
`
Also I was planning to keep `ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java` rule only for this PR and remove the other rule.
Some changes I wanted to do in the rule: 
1. As this rule currently checks for method name consist of `test`, but there are a lot of `setUpTest` that are set up methods and it was failing for those rules. So I wanted to delete the `test` check in name and = modify this rule as : 
2. Check if a method does not have `@isTest` annotation and if the method contains assert statements then throw the error:
pseudo code: 
```
if not have @isTest annotation and contains at-least 1 assert statement:
    throw error message 
```
Please let me know if this sounds good to you.

**Comment by adangel** [comment]
Hi @triandicAnt , you'll need to set JAVAFX_HOME to the base directory ("home") of javafx. For you it would be:
`JAVAFX_HOME=/Users/sudhansu/Downloads/javafx-sdk-11.0.2`
Maybe we should provide an example in our documentation?

> Some changes I wanted to do in the rule

Your proposed changes to "ApexUnitTestMethodShouldHaveIsTestAnnotationRule" sound fine. Just make sure to cover your found false-positives in test cases.

> remove the other rule (ApexUnitTestAssertStatement)

We actually came to an agreement to keep the rule, although it might not be best practice for everybody. We have the very same rule in Java: [JUnitAssertionsShouldIncludeMessage](https://pmd.github.io/pmd-6.12.0/pmd_rules_java_bestpractices.html#junitassertionsshouldincludemessage). So it makes sense to have this rule, just not enabled by default.

**Comment by triandicAnt** [comment]
@adangel thanks for the confirmation, I w'll doing the changes soon.

**Comment by triandicAnt** [comment]
@adangel I have made changes to this PR:
1. Update the name to `ApexAssertionsShouldIncludeMessageRule`.
2. Update the logic of `ApexUnitTestMethodShouldHaveIsTestAnnotationRule` to validate based on the assert statements inside the methos, rather than name containing `test`.
3. Added new test data to check false positives.
Please review and let me know incase of any changes.

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexAssertionsShouldIncludeMessageRule.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    protected Object checkForAssertStatement(ApexNode<?> node, Object data) {
+        // if System.assert has 1 argument, throw error
+        // if System.assertEquals/System.assertNotEquals has only 2 arguments, throw error
+        ASTMethodCallExpression assertMethodCall = (ASTMethodCallExpression) node;
+        if (assertMethodCall == null) {
+            return data;
+        }
+        if (ASSERT.equalsIgnoreCase(assertMethodCall.getFullMethodName())
```
Comment (adangel) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexAssertionsShouldIncludeMessageRule.java]: Please extract the method name first into a local variable:

```
String methodName = assertMethodCall.getFullMethodName();
```

It is used several times.

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexAssertionsShouldIncludeMessageRule.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                    "''{0}'' should have 2 parameters.",
+                    new Object[] { ASSERT });
+        } else if ((ASSERT_EQUALS.equalsIgnoreCase(assertMethodCall.getFullMethodName())
+                || ASSERT_NOT_EQUALS.equalsIgnoreCase(assertMethodCall.getFullMethodName()))
+                && assertMethodCall.jjtGetNumChildren() == 3) {
+            Object obj = ASSERT_EQUALS.equalsIgnoreCase(assertMethodCall.getFullMethodName())
+                    ? ASSERT_EQUALS : ASSERT_NOT_EQUALS;
+            addViolationWithMessage(data, node,
+                    "''{0}'' should have 3 parameters.",
+                    new Object[] { obj });
```
Comment (adangel) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexAssertionsShouldIncludeMessageRule.java]: Just use the method name - no need for an extra conditional statement to reverse engineer the method name again.

```suggestion
                    new Object[] { methodName });
```

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    private static final String ASSERT_NOT_EQUALS = "system.assertnotequals";
+
+    static {
+        ASSERT_METHODS.add(ASSERT);
+        ASSERT_METHODS.add(ASSERT_EQUALS);
+        ASSERT_METHODS.add(ASSERT_NOT_EQUALS);
+    }
+
+    public ApexUnitTestMethodShouldHaveIsTestAnnotationRule() {
+        addRuleChainVisit(ASTMethod.class);
```
Comment (adangel) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java]: As mentioned above: The base class `AbstractApexUnitTest` tries to apply this rule only for test classes. But adding a rule chain visit only for ASTMethod, makes this rule apply for any class.

Could this be a problem for this rule? It would report any method, that has a assert statement, regardless whether the method is contained in a test class or not...

If this is a problem, then we simply can't use the rule chain for this rule.

Are assert statements used outside of unit tests? If so, we should add a test case, where the class is a "normal" class (not a IsTest class) with a method, that contains a assert statement. This test case should not trigger the rule, since it is not a test class, and the method is therefore no test method.

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    @Override
+    public Object visit(final ASTMethod node, final Object data) {
+        // test methods should have @isTest annotation.
+        if (isTestMethodOrClass(node)) {
+            return data;
+        }
+        return checkForIsTestAnnotation(node, data);
+    }
+
+    private Object checkForIsTestAnnotation(final ApexNode<?> node, final Object data) {
```
Comment (adangel) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java]: This method actually checks for assert statements, not for IsTest annotations... So, please change the name.

Also, make sure, that the first argument is `ASTMethod` - then you don't need to cast later on. Also, you don't need to check for null.

File: pmd-apex/src/main/resources/rulesets/apex/ruleset.xml [hunk 7]
```diff
@@ -2,443 +2,50 @@
 <ruleset xmlns="http://pmd.sourceforge.net/ruleset/2.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" name="Default ruleset used by the CodeClimate Engine for Salesforce.com Apex" xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">
    <description>Default ruleset used by the Code Climate Engine for Salesforce.com Apex</description>
 
-   <!-- COMPLEXITY -->
```
Comment (adangel) [inline pmd-apex/src/main/resources/rulesets/apex/ruleset.xml]: I wouldn't have expected any change in this file... what is changed here?

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexAssertionsShouldIncludeMessageRule.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+ */
+public class ApexAssertionsShouldIncludeMessageRule extends AbstractApexUnitTestRule {
+
+    private static final Set<String> ASSERT_METHODS = new HashSet<>();
+    private static final String ASSERT = "System.assert";
+    private static final String ASSERT_EQUALS = "System.assertEquals";
+    private static final String ASSERT_NOT_EQUALS = "System.assertNotEquals";
+
+    static {
+        ASSERT_METHODS.add(ASSERT.toLowerCase(Locale.ROOT));
```
[the remark this answers is not in the document]
Comment (triandicAnt) [reply pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexAssertionsShouldIncludeMessageRule.java]: fixed

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexAssertionsShouldIncludeMessageRule.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    @Override
+    public Object visit(ASTMethodCallExpression node, Object data) {
+        if (!ASSERT_METHODS.contains(node.getFullMethodName().toLowerCase(Locale.ROOT))) {
+            return data;
+        }
+
+        return checkForAssertStatement(node, data);
+    }
+
+    protected Object checkForAssertStatement(ApexNode<?> node, Object data) {
```
[the remark this answers is not in the document]
Comment (triandicAnt) [reply pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexAssertionsShouldIncludeMessageRule.java]: fixed

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexAssertionsShouldIncludeMessageRule.java (hunk 4 above)
Comment (triandicAnt) [reply pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexAssertionsShouldIncludeMessageRule.java]: fixed

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java (hunk 5 above)
Comment (triandicAnt) [reply pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java]: AFAIK, It is not a good practice to use assert statements in non test classes. I will remove the class `addRuleChainVisit` from both of the rules.

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexAssertionsShouldIncludeMessageRule.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    private static final String ASSERT_NOT_EQUALS = "System.assertNotEquals";
+
+    static {
+        ASSERT_METHODS.add(ASSERT.toLowerCase(Locale.ROOT));
+        ASSERT_METHODS.add(ASSERT_EQUALS.toLowerCase(Locale.ROOT));
+        ASSERT_METHODS.add(ASSERT_NOT_EQUALS.toLowerCase(Locale.ROOT));
+    }
+
+    public ApexAssertionsShouldIncludeMessageRule() {
+        addRuleChainVisit(ASTMethodCallExpression.class);
```
[the remark this answers is not in the document]
Comment (triandicAnt) [reply pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexAssertionsShouldIncludeMessageRule.java]: removed `addRuleChainVisit ` and added test case to validate false positive cases.

File: pmd-apex/src/main/resources/rulesets/apex/ruleset.xml (hunk 7 above)
Comment (triandicAnt) [reply pmd-apex/src/main/resources/rulesets/apex/ruleset.xml]: this file was shown in merge conflict, so I pulled this from the upstream master.

**Comment by triandicAnt** [comment]
@adangel  Implemented review feedback. Added more test case to check false positive and removed `addRuleChainVisit` from both rule.

File: pmd-apex/src/main/resources/rulesets/apex/ruleset.xml (hunk 7 above)
Comment (adangel) [reply pmd-apex/src/main/resources/rulesets/apex/ruleset.xml]: This now adds an unrelated change into this PR... which is not desired.

Solving merge conflict can be done by either merging in the master branch into the PR-branch or rebasing the PR-branch against master. In both cases, you deal with the merge conflict, when you need to (at the time of merging/rebasing).

I'll undo the changes in this file before I merge this PR (since this PR should not change at all this file and thus completely avoids any merge conflicts in this file....).

**Comment by adangel** [comment]
@triandicAnt Thanks for you contribution! the 2 new rules will be part of PMD 6.13.0.

**Comment by triandicAnt** [comment]
@adangel  thank you!
